---
title: "4. Differential Gene Expression Analysis with DESeq2"
output:
  html_notebook: default
  html_document: default
  df_print: paged
  number_sections: yes
  toc: yes
---
<br>
CAGE data can also be used to assess expression of the cTSSs. Here, we will use the R-package DESeq2 [@love_moderated_2014]. Originated for RNA-seq data but can also handle similar data from other assaya types (such as CAGE data). The vignette and reference manual can be displayed by running the code below or can be found here: [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html). 
<br> <br>
Up to now, we have worked with just two samples. However, for a differential expression analysis you will need more samples such as replicates and/or more of the same group. To this end, we'll include two more samples to the mix as to follow the more standard work-flow and generate *p*-values. The two additional samples are from the same R package _ZebrafishDevelopmentalCAGE_. 
<br> <br>
The analysis will be on determing differentially expressed promoters in early stage expression vs later stage expression:

* __Early stages__:  zf_64cells & zf_512cells 
* __Late stages__:   zf_prim6 & zf_prim20


```{r, eval = FALSE}
browseVignettes("DESeq2")
```


## Summary and goals of this practical
<br>
_CAGEr_

* Prepare the right data format from CAGE data for DESEQ2
* Export the data from CAGEr

_DESeq2_

* Differential expression

_Follow up_

* Gene ontology



## 4.1 Exporting data from a CAGEset object
### Input for DESeq2
<br>
_DESeq2_ accepts matrices of read counts (count table) as input and this is exactly what we can export from _CAGEr_. The count data should be in the form of a matrix of integer values. The value in the _i-th_ row and the _j-th_ column of the matrix tells how many reads can be assigned to consensus cluster _i_ in sample _j_. The values in the matrix should be __raw counts__ of sequencing reads.

Consensus site (rownames) | Sample 1 | Sample 2 | Sample 3 | Sample 4   
-------------  | ------------- | ------------- | ------------- | -------------
1              | 0    | 1 | 44 | 60        
2              | 4 | 10 | 6 | 9       

> <small>!! "The DESeq2 model internally corrects for library size, so transformed or normalized values such as counts scaled by library size should not be used as input" </small>

#### The myCAGEset to be used in this practical
The code below was run to produce the consensus clusters for the four samples as this would take to long to do on the day. The one important (different step) is shown below in the code. Don't run the code (today).

```{r, eval=FALSE}
# packages
require(ZebrafishDevelopmentalCAGE)
require(CAGEr)

# load data
data(ZebrafishSamples)
samples <- as.character(ZebrafishSamples$sample)
myCAGEset <- importPublicData(source = "ZebrafishDevelopment", 
                              dataset = "ZebrafishCAGE", 
                              group = "development", 
                              sample = samples[c(3:4,11:12)] ) )

# CTSS tag count
ctss <- CTSStagCount(myCAGEset)

# !To keep using the raw counts in all downstream steps, the normalizeTagCount function of CAGEr should be used with the method set to "none". Note that normalizeTagCount function has to be applied to CAGEset object before moving to next steps.
normalizeTagCount(myCAGEset, method = "none")

# Clustering of CTSS
clusterCTSS(object = myCAGEset, 
            threshold = 1, 
            thresholdIsTpm = TRUE, 
            nrPassThreshold = 1, 
            method = "distclu", 
            maxDist = 20, 
            removeSingletons = TRUE, 
            keepSingletonsAbove = 5)

# cumulative distribution and quantile positions
cumulativeCTSSdistribution(myCAGEset, clusters = "tagClusters")
quantilePositions(myCAGEset, 
                  clusters = "tagClusters", 
                  qLow = 0.1, 
                  qUp = 0.9)

# aggregate the clusters across the samples:
aggregateTagClusters(myCAGEset, 
                     tpmThreshold = 5, 
                     qLow = 0.1, 
                     qUp = 0.9, 
                     maxDist = 100)

# save the object
save(myCAGEset, file = "../Data/provided/AggregatedTagClus_0109_4Samples.RData")
```
<br>

#### Extract the raw count table 

We want to compare transcriptional activity across these samples, therefore we will use consensus clusters for downstream analysis. This is provided to you in a .RData file using the code above. 
```{r,message = FALSE}
# package
require(CAGEr)

# the file
load("../Data/provided/AggregatedTagClus_0109_4Samples.RData")
```


<br> <br> 
First we extract the count matrix per consensus cluster for each sample:
```{r, message = FALSE}
count.df <- data.frame(consensusClustersTpm(myCAGEset))
```
Look at the dataframe to see what is there:
```{r}
head(count.df)
```

As you can see, the consensus coordinates are not automatically extracted. We will have to do that seperately with the **consensusClusters( )** function:

```{r}
consensus.info <- consensusClusters(myCAGEset)
head(consensus.info)
```
<br>
Do these two have the same number of rows? 
```{r}
nrow(count.df) == nrow(consensus.info)
```
Yes! That is because the order of values will be the same from the myCAGEset. So all we have to do is combine the two. To be sure let's create a consensus identifyer too while we are at it:

```{r}
# create identifiers to link back
consensus.info$cons_clus_id <- paste("cid_",1:nrow(consensus.info), sep = "")
rownames(count.df) <- consensus.info$cons_clus_id

# the order is the same so we can easily use "cbind"
count.consensus.info <- cbind(consensus.info[,-1], count.df)
write.table(count.consensus.info, 
            "../Data/intermediate/CountTable_Consensus_4Samples.txt", 
            col.names = TRUE, 
            row.names = FALSE, 
            sep = "\t", 
            quote = FALSE )

# remove unnecessary files 
rm(myCAGEset, consensus.info)
```


<br>

## 4.2 Differential Gene Expression Analysis
<br>

#### Creating DESeqDataSet object
So we have our count table and the first thing to do is to make a _DESeqDataSet object_. 
We will also add the "formula" which is the design of the analysis downstream (a linear model: ~ condition). In this example we are only using the variable (early vs late).


```{r}
# DESeq2 expects a matrix of count table:
count.matrix <- as.matrix(count.df) 
```
If from previously saved file:
```{r}
count.consensus.info <- read.table("../Data/intermediate/CountTable_Consensus_4Samples.txt", 
                        header = TRUE, 
                        sep = "\t")
count.matrix <- as.matrix(count.consensus.info[,7:10])
```
The condition for the analysis in a dataframe format:
```{r}
samples <- colnames(count.matrix)
info.df <- data.frame(condition = factor(x = c("early", "early", "late", "late"), 
                                    levels = c("early","late")), 
                      row.names = samples)

```
> If you want to add covariates in the model (i.e. batch) these should be extra columns in the info.df and be given in the model like this: ~ column name covariate + variable. Keep in mind that the variable needs to be the last in the model. 

The reason for specifying the levels of the factor is because by default, R will handle factor levels based on alphabetical order. It is good practice to get into, especially in analyses where it matters which level you want to compare against. Thus identifying correctly up- or down-regulated genes in this case.
<br> <br>
Let's create a DESeqDataSet object that stores the count matrix and design:

```{r, message = FALSE}
require(DESeq2)
dds <- DESeqDataSetFromMatrix(countData = count.matrix, 
                              colData = info.df, 
                              design = ~ condition)
```
<br>

#### Differential expression analysis
The standard differential expression analysis steps are wrapped into a single function, **DESeq( )**. Results tables are generated using the function __results( )__, which extracts a results table with log2 fold changes, pvalues and adjusted pvalues. The text, condition treated vs untreated, tells you that the estimates are of the logarithmic fold change log2 (treated/untreated). 

```{r, message = FALSE}
# the analysis
dds <- DESeq(dds)
# results
res <- results(dds)
```

<br> 
We probably want to know what are the lowest adjusted p-values, which genes they represent (if any), and gene ontology for patterns in the data.
<br>
First, we are adding the genomic coordinates of the consensus clusters. These are still stored in:
```{r, eval=FALSE}
count.consensus.info
``` 
Again, the order has not changed so let's add it to the results and then reorder the results according to lowest p-value.

```{r}
res.info <- cbind(count.consensus.info[,c(1:4,6)], data.frame(res@listData))
res.info <- res.info[order(res.info$padj),]
head(res.info)
```
An easy summary of the results:
```{r}
summary(res)  # orignal S4 output of DESeq
```

The amount consensusclusters differentially expressed (padj < 0.05):

```{r}
sum(res$padj < 0.05, na.rm=TRUE)
```

Save the results in a table in the Intermediate directory:
```{r}
write.table(res.info, 
            "../Data/intermediate/DiffExpression_Consensus_4Samples.txt", 
            col.names = TRUE, 
            row.names = FALSE, 
            quote = FALSE, 
            sep = "\t")
```



# 4.3Gene ontology
<br>

#### GREAT



Let's start loading in the TxDb for our samples that can be found in the _provided data directory_. 


Load in the the txdb:
```{r}
require(GenomicFeatures)
require(AnnotationDbi)
txdb <- loadDb("../Data/provided/txdb_DanRer7.sqlite")
```


To use the useful features of GenomicRanges we will have to convert our differential expression result to a GRanges object too:

```{r}
# packages
require(BSgenome.Drerio.UCSC.danRer7)

total = res.info 
gr <- GRanges(seqnames = total$chr,
                ranges = IRanges(start = total$start,end = total$end),
                strand = total$strand,
                cons_clus_id = total$clus.gene, # the identifier
                seqlengths = seqlengths(Drerio))
```


```{r}
require(biomaRt)

# older version of ensembl 79
ensembl =useMart('ENSEMBL_MART_ENSEMBL', host = 'mar2015.archive.ensembl.org', dataset='drerio_gene_ensembl')



test <- getBM(
    attributes=c('hgnc_symbol'), 
    filters = c('chromosome_name', 'start', 'end' ), 
    values = with(res.info, list(gsub("chr", "", as.character(chr)), start, end)),
    mart = ensembl )

```


# References